File: python/day_11/part_1.py
def solve(input: str) -> int:
    lines = [line.strip() for line in input.strip().split("\n") if line.strip()]

    graph = {}
    for line in lines:
        parts = line.split(":")
        device = parts[0].strip()
        outputs = parts[1].strip().split()
        graph[device] = outputs

    memo = {}

    def count_paths(node):
        if node == "out":
            return 1

        if node in memo:
            return memo[node]

        if node not in graph:
            return 0

        total = 0
        for neighbor in graph[node]:
            total += count_paths(neighbor)

        memo[node] = total
        return total

    return count_paths("you")


if __name__ == "__main__":
    with open("inputs/day_11.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: typescript/day_11/part_1.ts
import { readFileSync } from 'fs';

function solve(input: string): number {
    const lines: string[] = [];
    for (const line of input.trim().split('\n')) {
        const trimmed = line.trim();
        if (trimmed) {
            lines.push(trimmed);
        }
    }

    const graph = new Map<string, string[]>();
    for (const line of lines) {
        const parts = line.split(':');
        const device = parts[0].trim();
        const outputs = parts[1].trim().split(/\s+/);
        graph.set(device, outputs);
    }

    const memo = new Map<string, number>();

    function count_paths(node: string): number {
        if (node === 'out') {
            return 1;
        }

        if (memo.has(node)) {
            return memo.get(node)!;
        }

        if (!graph.has(node)) {
            return 0;
        }

        let total = 0;
        for (const neighbor of graph.get(node)!) {
            total += count_paths(neighbor);
        }

        memo.set(node, total);
        return total;
    }

    return count_paths('you');
}

const input = readFileSync('../inputs/day_11.txt', 'utf-8');
const result = solve(input);
console.log(result);
